litgpt/api.py

```diff
@@ -213,19 +216,55 @@ def generate(
 
                 For more details, see https://arxiv.org/abs/1904.09751
                 or https://huyenchip.com/2024/01/16/sampling.html#top_p
+            return_as_token_ids: If True, returns the token IDs as a torch.Tensor. Otherwise, returns the decoded text as a string.
+            stream: If True, returns a generator that yields tokens as they are generated.
+                At the moment, this setting is slower and may use more memory than the non-streaming version.
+                We plan to resolve this in the future.
         """
         prompt = self.prompt_style.apply(prompt)
         input_ids = self.preprocessor.tokenizer.encode(prompt)
         prompt_length = input_ids.size(0)
         max_returned_tokens = prompt_length + max_new_tokens
 
-        if self.model.mask_cache is not None:
+        # Create, clear or grow the kv cache if necessary.
+        kvcache_uninitialized = self.model.mask_cache is None
+        
+        if not kvcache_uninitialized:
             self.model.clear_kv_cache()
+    
+        prev_size = self.model.mask_cache.size(-1) if not kvcache_uninitialized else None
+        max_model_supported = self.model.max_seq_length
+
+        if max_seq_length == 'dynamic':
+            if max_returned_tokens > max_model_supported:
+                raise ValueError(
+                        f"Cannot generate a response with {max_returned_tokens} tokens.\n"
+                        f"This model has a maximum context length of {max_model_supported} tokens.\n"
+                        f"The prompt contains {prompt_length} tokens, leaving {max_model_supported - prompt_length} for the response, which is not enough."
+                    )
+            if kvcache_uninitialized or prev_size < max_returned_tokens:
+                self.model.set_kv_cache(batch_size=1, max_seq_length=max_returned_tokens, device=self.fabric.device)
+        elif max_seq_length == 'max_model_supported':
+            if kvcache_uninitialized or prev_size != max_model_supported:
```

Comment:
I wonder when would 

```python
prev_size != max_model_supported
```

in 

```python
        elif max_seq_length == 'max_model_supported':
             if kvcache_uninitialized or prev_size != max_model_supported:
```

be `True`, and if there are cases where this is true, wouldn't that trigger recompilation? Is this desired?